### PR TITLE
Implement normal set literal

### PR DESCRIPTION
**Pull request by GreasySlug**
Implementation is simple.
If there are duplicate element(s), they are not pushed.

- [ ] Completed implement normal set
    - [x] parse
    - [x] hir
    - [ ]  lower
        - [x] add `Set(T, N)`/`Set!(T, !N)` type (almost same as `Array!(T, N)`/`Array!(T, !N)`)
        - [x] register `Set`, `Set!` definitions to the builtin `Context`
        - [x] check if objects implement `Eq`
        - [ ] (optional) warn of any duplicate elements
        - [x] tyvar
        - [x] link
    - [x] effectcheck
    - [x] memcheck
    - [x] codegen
 
 Additional
- [ ] comparing 
- [ ] set operations
    - [ ] and
    - [ ] or
    - [ ] in
    - [ ] not

@mtshiba

### CODE REVIEW

**Comment by mtshiba** [comment]
The difference between a Set and an Array is that duplicate elements are removed. Otherwise, it is almost the same as Array, so please refer to the Array implementation.

**Comment by GreasySlug** [comment]
Why was the attribute added?
I understand that it will probably be used, but where will it be used?

**Comment by mtshiba** [comment]
Come to think of it, `Set`, unlike `Array`, cannot store incomparable objects. If it is difficult to implement the checking, I will do it.

File: compiler/erg_compiler/lower.rs [hunk 1]
```diff
@@ -368,13 +368,28 @@ impl ASTLowerer {
                     ),
                 )));
             }
-            new_set.push(elem);
+            if !new_set.insert(elem.clone()) {
+                self.warns.push(LowerWarning::syntax_error(
```
Comment (GreasySlug) [inline compiler/erg_compiler/lower.rs:372]: Pushing sytax_error() to use warn
Is this usage correct?

**Comment by GreasySlug** [comment]
cargo r

```sh
>>> a = {"hi", "there", "hi", "world", "world", "hi"} 

>>> a
{'hi', 'there', 'world'}
>>> c = !{1, 2, 1, 2} 

>>> c
{1, 2}
```

cargo lwr

```sh
>>> a = {1, 2, 1, 2}
::a(: {2, 1, }) =
    {
        2 (: {2, })
        1 (: {1, })
    }(: {2, 1, }) # The type is not Set type
```

**Comment by mtshiba** [comment]
> cargo r
> 
> ```shell
> >>> a = {"hi", "there", "hi", "world", "world", "hi"} 
> 
> >>> a
> {'hi', 'there', 'world'}
> >>> c = !{1, 2, 1, 2} 
> 
> >>> c
> {1, 2}
> ```
> 
> cargo lwr
> 
> ```shell
> >>> a = {1, 2, 1, 2}
> ::a(: {2, 1, }) =
>     {
>         2 (: {2, })
>         1 (: {1, })
>     }(: {2, 1, }) # The type is not Set type
> ```

**Comment by mtshiba** [comment]
> cargo r
> 
> ```shell
> >>> a = {"hi", "there", "hi", "world", "world", "hi"} 
> 
> >>> a
> {'hi', 'there', 'world'}
> >>> c = !{1, 2, 1, 2} 
> 
> >>> c
> {1, 2}
> ```
> 
> cargo lwr
> 
> ```shell
> >>> a = {1, 2, 1, 2}
> ::a(: {2, 1, }) =
>     {
>         2 (: {2, })
>         1 (: {1, })
>     }(: {2, 1, }) # The type is not Set type
> ```

In Array:

https://github.com/erg-lang/erg/blob/f8fa9929301d488622b66194e352ca31ca0c54b4/compiler/erg_compiler/hir.rs#L618

**Comment by GreasySlug** [comment]
Build error could not be avoided
Could you give me any advice as to how to determine if `Eq` is implemented or not?

**Comment by mtshiba** [comment]
Normally, if you implement `PartialEq`, you can just implement `Eq` without thinking.

Structures involving floats should not implement Eq, though, since floats do not implement Eq.

**Comment by mtshiba** [comment]
Maybe the cache is doing something bad. I built the source and it is fine, just ignore it.

**Comment by mtshiba** [comment]
You can see the error log from `Checks`.

スクリーンショット 2022-10-06 115042

**Comment by mtshiba** [comment]
Since the basic features seems to have been implemented, I'll merge it for now, and/or, etc. will be implemented later.

**Pull request merged**